<html>
<head>
  <meta http-equiv="content-type" content="text/html; charset=UTF-8">
  <title>AliView - Alignment Viewer and Editor</title>
  <link rel="shortcut icon" href="alignment_ico_64x64.png">
</head>

<body>
<h1>aliview</h1>

<p>AliView is yet another alignment viewer and editor, but this is probably one
of the fastest and most intuitive to use, not so bloated and hopefully in your
likings...</p>
<img style="width: 383px;" alt="" src="images/prog1.png"><br>


<p>The general idea when designing this program has always been speed and
usability, all new functions were always optimized so they did not affect the
general performance and capability to work with unlimited file sizes. </p>

<p>A need to easily sort, view, remove, edit and merge sequences from large
transcriptome datasets initiated the work with the program.</p>
- edit (manually)<br>
- align, add and align automatically (with muscle or mafft or any other aligner
of your choise)<br>
- find degenerate primers in conserved regions in an alignment of mixed
species<br>
- handles fasta, nexus or phylip-format (unlimited file sizes)<br>
- on the fly translation of nucleotides to amino acids<br>
- various visual cues to highligt consensus characters or characters deviating
from consensus<br>
- simple copy/paste/drop/remove of sequences/files<br>
- a very simple to use "extenal interface" that lets you invoke other of your
favourite programs (you could for example automatically have the alignment sent
to FastTree and then automatically opened in FigTree).<br>
<br>
The program is developed at department of Systematic Biology (Uppsala
University) so there is probably a predominance in functionality supporting
someone working with phylogenies.<br>


<p><a href="downloads/">download here (Linux, Mac OSX, Windows)</a></p>
The program is released under Open Source Software License 'GNU General Public
License, version 3.0 (GPLv3)' <br>
-source code is available at GitHub: <a
href="https://github.com/AliView/AliView">https://github.com/AliView/AliView</a><br>


<h2>contact: </h2>

<p>Anders Larsson,<br>
Dep. of Systematic Biology<br>
Uppsala University<br>
Sweden<br>
email: anders.larsson [at] ebc.uu.se<br>
Phone: +46 (0)18 471 2932<br>
Fax: +46 (0)18 471 6457 </p>
<br>
<br>


<h2>Highlights:</h2>

<p>Don't miss the mouse-wheel zoom out functinality to see the whole alignment
at once: </p>

<p style="font-weight: bold;"><img style="width: 500px;" alt=""
src="images/prog_zoom_out.png"></p>
<br>


<h2><a name="DOWNLOAD">Install Instructions and Downloads</a></h2>

<h3>Linux (all versions)  <a href="downloads/">download here</a></h3>
aliview.install.run<br>
<br>
(this is a executable archive that will copy the files to the /usr/home/bin/
and /usr/share)<br>
on some systems you need to change the execution rights of the install file:<br>
<br>
<code>chmod +x aliview.install.run <br>
sudo ./aliview.install.run </code><br>
<br>
(most likely you need to sudo the installation)<br>
<br>
start the program with aliview (this is a sh script that will execute java -jar
aliview.jar)<br>
<br>
<b>Linux (alternative)</b><br>
aliview.tgz is an archive including the program files<br>
<br>


<h3>Mac OS X (all versions)  <a href="downloads/">download here</a> </h3>
AliView-app.zip<br>
<br>
the program is distributed as a Application that only need to be dropped in
your App-folder from (sometimes you need to unpack it manually)<br>
<br>
<b>NOTE:</b> on Mavericks you might get an error message <b>”This
application  is damaged and can't be opened. You should move it to the
Trash.”</b> - The file is actually not damaged, it is just a bad error
message. There is a settings in the ”Gatekeeper” that you need to change to
be able to download software from other places than the App-store.<br>
How to solve:<br>
<a
href="http://answers.uchicago.edu/page.php?id=25481">http://answers.uchicago.edu/page.php?id=25481
</a><br>
<br>
<br>


<h3>Windows (all versions)  <a href="downloads/">download here</a></h3>
AliView-Setup.exe (There are also an alternative executable without installer
and one version packaged in .jar file)<br>
<br>
Run the installer, the program will default installation to directory
C:\Program Files (x86)\AliView 

<p><strong>NOTE:</strong> some virus scanners might block the installation and
need to be temporarily disabled.<br>
<br>
<img style="width: 371px; height: 288px;" alt=""
src="images/windows-install.png"><br>
<br>
</p>

<p><br>
<br>
</p>

<h1><a name="BUG_OR_FEATURE" id="BUG_OR_FEATURE">Report a bug or feature
request:</a></h1>
Two ways: <br>
<br>
1. Report issue at AliView GitHub Issues tracker: <a
href="https://github.com/AliView/AliView/issues/">https://github.com/AliView/AliView/issues/</a>
<br>
or <br>
2. Send email to: anders.larsson@ebc.uu.se <br>


<p></p>

<p></p>

<p></p>

<h1>Feature-list:</h1>

<p>- fast and light-weight <br>
- simple navigation, mouse-wheel zoom out unlimited to see whole alignment, and
zoom in on place of interest. <br>
- various visual cues to highlight consensus characters or characters deviating
from consensus or characters deviating from a selected "trace"-sequence <br>
- edit sequences/alignment (manually), insert, delete, change, move, rename
(with keyboard or mouse) <br>
- align, add and align automatically with MUSCLE (included) or MAFFT, or any
other aligner of your choice <br>
- define aligner program presets (different parameters, different software) <br>
- align new sequences to existing or realign all <br>
- realign a selected block <br>
- realign nucleotides as translated amino-acids <br>
- delete vertical gaps <br>
- undo/redo <br>
- find degenerate primers in conserved regions in an alignment of mixed species
<br>
- open and save in FASTA-, NEXUS- or PHYLIP-format (unlimited file sizes) <br>
- print (current view) <br>
- export whole alignment as an image file (png-format). <br>
- copy selection (as fasta-sequences or just characters) <br>
- paste sequences (both as sequences in clipboard or filenames in clipboard)
<br>
- add more sequences from file <br>
- merge two sequences - and calculate consensus if the overlap is not perfect
<br>
- translate (view) nucleotide sequences as amino-acid sequences <br>
- save translated alignment <br>
- read and preserve Codonpos, Charset and Excludes (Nexus-specification) <br>
- change Codonpos for selected regions (Nexus-specification) <br>
- drag-drop/remove of sequences/files <br>
- move sequences to top/bottom with key-press <br>
- a very simple to use "extenal interface" that lets you invoke other of your
favorite programs (you could for example automatically have the alignment sent
to FastTree and then automatically opened in FigTree). <br>
- search function that finds patterns across gaps and follow IUPAC codes. <br>
- search multiple sequence names at once (stored in clipboard) <br>
- reverse complement/reverse/complement sequences or whole alignment <br>
- different color schemes including ClustalX <br>
- sort sequences by name <br>
</p>

<p></p>

<p></p>

<h1><a name="TOP_HELP" id="TOP_HELP">Help section:</a></h1>

<p><br>
</p>

<h2>File menu:</h2>

<p><strong>Open File</strong>: AliView will open files in Fasta, Nexus and Phylip format.</p>

<p><strong>Save as Nexus simplified names</strong>: Unorthodox characters (-
\?./|\,&amp;) in Sequence names are replaced by "_"</p>

<p><strong>Save as Codonpos Nexus (nucleotide dataset only):</strong> All pos 1
nucleotides are saved as one charset in begining of alignment, thereafter all 2
pos nucleotides as one charset, then all 3 pos and finally all non-coding
(0-pos). </p>

<p><strong>Print:</strong> Apart from printing to paper, it is often possible
in the settings to "print to file", this way you can get post script (ps) files
of the alignment.</p>

<p><strong>Export alignment as image:</strong> This will save the entire
alignment as an .png image file</p>

<p><strong>Save selection as Fasta:</strong> Only the selected sequences, or
selected parts of sequences will be included in the .fasta file - a very simple
way of pruning away things from your alignment.</p>

<p><strong>Save Translated alignment as Amino acids (nucleotide dataset
only):</strong> Program will translate nucleotides into AminoAcids (taking into
acound user specified edits in codon positions when translating)</p>

<p><strong>Start debug:</strong> Prints more messages</p>

<p><strong>Show message log:</strong> Displays program messages and errors in a
message window</p>

<p></p>

<h2>Edit:</h2>

<p>When opening very large files and they are indexed, then the editing capabilities are very limited - but if you need to edit it might be possible if you increase the Memory resources for AliViev <a href=#memory_settings>see here</a>.

<p><strong>Undo/Redo:</strong> AliView has undo/redo functionality (except from when large indexed files are read from file). Depending on the <a href=#memory_settings>Memory settings</a>, the number of possible undo/redo steps change.</p>

<p><strong>Copy selection as Fasta:</strong> The selected sequences/characters are copied into clipboard as a Fasta sequences (including both name and sequence).</p>

<p><strong>Copy selection as characters:</strong> The selected sequences/characters are copied into clipboard as characters only - Not including the names of the sequences where characters come from.</p>

<p><strong>Paste (fasta sequences):</strong> Sequence(s) will be pasted at top of alignment if clipboard contains <strong>either</strong> File containing fasta sequences <strong>or</strong> text in fasta format, e.g.:<br>
>Name_of_seq<br>
TTGGGCATG<br>
</p>

<p><strong>Add sequences from file:</strong> Add sequence(s) from a file to be pasted at top of alignment (FASTA,PHYLIP or NEXUS file format).</p>

<p><strong>Edit mode:</strong> It is not possible to edit alignment by accident unless you activate "Edit Mode".</p>

<p><strong>Edit mode:</strong> It is not possible to edit alignment by accident unless you activate "Edit Mode".</p>

<p><strong>Clear selected bases:</strong> Changes selected nucleotides/AminoAcids into GAP.</p>

<p><strong>Delete selected</strong> Removes selected sequences/nucleotides/AminoAcids from alignment.</p>

<p><strong>Delere selected bases:</strong> Removes selected nucleotides/AminoAcids from sequences/alignment.</p>

<p><strong>Delete vertical gaps:</strong> If there are column(s) in the alignment only containing GAP these positions are removed.</p>

<p><strong>Delete all gaps in all sequences:</strong> Degap the whole alignment.</p>

<p><strong>Trim sequences/alignment:</strong> If there are column(s) in the beginning or at the end of the alignment containing GAP only, then these positions are removed.</p>

<p><strong>Excludes (Nexus specification)</strong> can be useful if you have an alignment with sections that you don't want to remove permanently - but sometimes when you make an analysis. Then you can save the alignment with the "Excluded" sections removed and use that alignment for the analysis.</p>

<p><strong>Delete excluded bases:</strong> Removes sections of the alignment are marked as "Excludes" (Nexus specification). See also avove under Excludes (Nexus specification).</p>

<p><strong>Find </strong> Finds in name or nucleotides/aminoacids - searches across gaps and obey IUPAC codes.</p>

<p><strong>Find sequence names from clipboard</strong> Finds and selects secuence(s) matching names that are srored in clipbord as list.<br>
example list in clipboard:<br>
Woodsia<br>
Athyrium<br>
Dryopteris
</p>

<p><strong>Merge two selected sequences:</strong></p>

<h2>Selection:</h2>

<p><strong>Add selection to Excludes:</strong> Marks sections of the alignment as "Excludes" (Nexus specification). See also avove under Excludes (Nexus specification).</p>

<p><strong>Remove selection from Excludes:</strong> Removes selected sections of the alignment from "Excludes" (if they are included already). See also avove under Excludes (Nexus specification).</p>



<p><strong></strong></p>

<p></p>

<p></p>

<p></p>

<p></p>

<p></p>

<p></p>

<h2><a name="large_file">Load more sequences from file</a></h2>

<p>If you have opened a large file and all sequences were not indexed at once, this menu will show up and you can select other parts of the file to index. You can also change the number of sequences to index at once in the Program preferences, <a href=#General_settings>General settings</a></p>


<p></p>

<h2><a name="memory_settings">Memory settings</a></h2>

<p>If you want AliView to read larger alignments in memory and not from file
(this allows for more editing capabilities), then you can change the maximum
memory settings for the program.</p>

<p><b>Mac OS X</b> </p>

<p>Go to Applications in Finder --&gt;Left click on specific application
AliView --&gt; Show Package Content --&gt; Contents --&gt; Then open the file
"Info.Plist" in a text-editor and change the parameter: &lt;string&gt;-Xmx512m
-Xms128m&lt;/string&gt; to something different (for example 2GB=2048M):<br>
&lt;string&gt;-Xmx2048m -Xms128m&lt;/string&gt;</p>

<p><b>Linux</b> </p>

<p>/usr/bin/aliview <br>
open this file in text-editor and change the parameter -Xmx1024M (default
setting = 1024M memory)</p>

<p><b>Windows</b> </p>

<p>In the installation folder of AliView (default: "c:\Program Files\AliView\")
open the file "AliView.l4j.ini" in a text editor and change the setting:
-Xmx1024m to something you prefer (for example 2GB=2048M) -Xmx2048m</p>

<p></p>

<h2><a name="General_settings">General setting</a></h2>
<img style="width: 522px; height: 361px;" alt=""
src="images/general_settings.png"><br>
<br>
<b>Reverse horizontal mouse-wheel (touchpad) movement <br>
Reverse vertical mouse-wheel (touchpad) movement </b> <br>
(if the direction if movement is driving you crazy - adjust it here).<br>
<br>
<b>Horizontal scroll speed <br>
Verticall scroll speed<br>
</b> (on some systems the movement is to fast/slow) – you can adjust it
here.<br>
<br>
<b>Override default program font size</b> <br>
(on some computers with extreme resolution it might be hard to read text with
default settings)<br>
<br>
When working with large files being read straight from disk:<br>
<b>Number of sequences to Index at a time in very large files</b><br>
(you can change the number of sequences being indexed at once – the rest of
the sequences in the file is available by selecting the menu-alternative
”Load more sequences” (this menu is only showing up when the file is parted
into ”pages”)<br>
<br>
<b>Number of sequences to count when calculating ClustalX consensus on large
files</b><br>
(it would take very long time to calculate consensus considering every position
in a large file)<br>


<h2><a name="Find_Primer_settings">Find Primer settings</a></h2>
<img alt="" style="width: 600px;" src="images/primersettings.jpg"><br>
<br>
<br>


<h2><a name="ALIGNER_SETTINS_ALL">Program to use for "Align all"
commands</a></h2>
<img style="width: 600px;" alt="" src="images/align_all_settings.png"><br>
<br>
In the <b>Program</b> field you enter the program you want to execute when
using the alignment functions in AliView.<br>
<br>
There are default settings for Muscle and Mafft included in AliView and they
look different for different operating systems (Windows / Linux / Mac OSX).<br>
<br>
<strong>Note 1:</strong>It is essential that filen names that contains spaces
or unorthodox characters are specified within citation marks <b>”</b> <br>
On Windows for example: ”c:\Program Files\mafft-win\mafft.bat”<br>
On Mac OSXf or example: ”FigTree 1.4.0.App”<br>
<br>
<strong>Note 2: </strong>When calling MAFFT on Windows computers you are
actually starting the program cmd.exe and instead including mafft program path
as a parameter (compare for example with the muscle.exe program and
command).<br>


<h2><a name="ALIGNER_SETTINS_ADD">Program to use when adding and aligning in
new sequences ("Align add" commands)</a></h2>
<img style="width: 600px;" alt="" src="images/align_add_settings.png"><br>
<br>
<br>
For explanation on how to fill in the fields see above "Program to use for
"Align all" commands"<br>
<br>


<h2><a name="EXTERNAL_COMMANDS">External commands - Send the alignment to other
programs</a></h2>
<img style="width: 600px;" alt="" src="images/external_program_settings.png">
<br>


<p>You can send the active alignment from AliView as a command parameter to
other programs and create a simple pipeline of commands executed from a
Menu-button in AliView. </p>
Each line is executed as a separate command waiting for the previous to have
finished, as an example is a simple combination of commands that are taking the
current alignment and running it in FastTree to create a MaximumLikelihood Tree
and finaly opening the resulting tree in FigTree.<br>
<br>
Special Parameters:<br>
CURRENT_ALIGNMENT_FASTA =The current alignment as a fasta file)<br>
CURRENT_SELECTION_AS_FASTA = The selected sequences as a fasta file<br>
TEMP_OUT_FILE = An empty temporary file created by AliView in the /temp/ folder
of the computer<br>


<h3>examples:</h3>
Here are examples


<h2>Drag and Drop</h2>

<p>
If you drop a file on the program window it will open a new window with the alignment
<br>
<strong>Note:</strong> If you drop a sequence with "Shift"-button pressed - the sequences in the file will be added to the currently loaded alignment.
</p>
Drag and Drop



</body>
</html>



